pkg/controller/scyllacluster/resource.go

```diff
@@ -625,6 +637,121 @@ func MakePodDisruptionBudget(c *scyllav1.ScyllaCluster) *v1beta1.PodDisruptionBu
 	}
 }
 
+func MakeIngresses(c *scyllav1.ScyllaCluster, services map[string]*corev1.Service) []*networkingv1.Ingress {
+	// Don't create Ingresses if cluster isn't exposed.
+	if c.Spec.ExposeOptions == nil {
+		return nil
+	}
+
+	type params struct {
+		backendName    string
+		portName       string
+		ingressOptions *scyllav1.IngressOptions
+	}
+	var ingressParams []params
+
+	if c.Spec.ExposeOptions.CQL != nil && isIngressEnabled(c.Spec.ExposeOptions.CQL.Ingress) {
+		ingressParams = append(ingressParams, params{
+			backendName:    "cql",
+			portName:       portNameCQLSSL,
+			ingressOptions: c.Spec.ExposeOptions.CQL.Ingress,
+		})
+	}
+
+	var ingresses []*networkingv1.Ingress
+
+	for _, ip := range ingressParams {
+		for _, service := range services {
+			var hosts []string
+			labels := naming.ClusterLabels(c)
+
+			switch naming.ScyllaServiceType(service.Labels[naming.ScyllaServiceTypeLabel]) {
+			case naming.ScyllaServiceTypeIdentity:
+				for _, domain := range c.Spec.DNSDomains {
+					hosts = append(hosts, fmt.Sprintf("%s.%s.%s", naming.ScyllaIngressSubdomainAny, ip.backendName, domain))
+				}
+				labels[naming.ScyllaIngressTypeLabel] = string(naming.ScyllaIngressTypeAnyNode)
+
+			case naming.ScyllaServiceTypeMember:
+				hostID, ok := service.Annotations[naming.HostIDAnnotation]
+				if !ok {
+					klog.V(4).Infof("Service %q is missing HostID annotation, postponing Ingress creation until it's available", naming.ObjRef(service))
+					continue
+				}
+
+				if len(hostID) == 0 {
+					klog.Warningf("Service %q has empty HostID annotation, postponing Ingress creation until it's fixed", naming.ObjRef(service))
```

Comment:
as this is an unrecoverable error, likely caused by the issuer or a bug - we should make the user aware using an event

Reply:
I guess our sidecar controller should always override it so as long as it work this would only be intermittent, so warning may be fine

Reply:
it's syncing what Scylla API is returining, so if label value doesn't match it's going to be overwritten

Reply:
If this is a nasty user playing with it, the controller shouldn't error. Also error will just make the case an infinite retry with backoff, yet the resource  won't result in a different codepath.

The option I see is keeping the warning, or sending an event. As it is supposed to be reconciled back by the sidecar, I'd lean towards the warning, sorry for the false alarm here.

Reply:
> Also error will just make the case an infinite retry with backoff, yet the resource won't result in a different codepath.

It won't be infinite, because sidecar controller doesn't use hash annotation to check whether resource match it's expectation, it will react on every service label change (user interrupting) and it will override and fix the value with what Scylla returns. 
Leaving warning would not propagate error to upper level where reconcilation status is calculated, and without one required resource, cluster should be marked as degraded. We don't mark that for other errors as well, but that's outside of scope of this PR.

Reply:
Let's think about controllers alone, assume only the scyllacluster controller running and the sidecar one being broken. ScyllaCluster object has a broken annotation and the scyllacluster controller returns error. That means in a bit scyllacluster controller is gonna try again on exactly the same object, and again with the backoff policy... If someone fixes the annotation, you don't need the retries because you'll get requeued on changes.

Reply:
Without error, controller sync loop will succeed, properly working Ingress will be removed if it was created before interruption, and status will reflect that cluster is operational even when one of the required resources isn't created and user can't connect to all nodes in it. 

Brought back warning, even though I consider error a better choice overall here.

Reply:
> Without error, controller sync loop will succeed, properly working Ingress will be removed if it was created before interruption

that's actually a good thing, the ingress no longer matches the desired state and is a relic of the past. whether it's still valid or not is just a guessing game - e.g. what if the hostid changed but there is none projected yet, so implicitly assuming it's still good may not be a good idea

Reply:
and thanks ;)

Reply:
and thanks for the update ;)

(the isolated case shows that a controller shouldn't requeue based on the state of the resource, only on external actions)

Reply:
> is just a guessing game

Exactly, and controller guessed it's better to act (remove ingress) instead doing nothing because it ecountered undefined behavior. Imo it's wrong, because even when state is invalid, node might be functional and even though it cascades to application interruptions. 
Doing nothing slightly increases load on apiserver which isn't a big thing compared to possibly loosing access to precious (hopefully replicated) data.

Reply:
well, I think it goes the other way :) if the spec doesn't say there should be a managed ingress and there is one, it shouldn't be there.